### ISSUE TITLE
The unnecessary limitations of the Cache.put API leads to users inventing work-arounds 

### ISSUE THREAD

**Issue by expenses** [w3c/ServiceWorker#1643]
Step 4. of https://w3c.github.io/ServiceWorker/#cache-put says:

> If innerRequest’s [url](https://fetch.spec.whatwg.org/#concept-request-url)'s [scheme](https://url.spec.whatwg.org/#concept-url-scheme) is not one of "http" and "https", or innerRequest’s [method](https://fetch.spec.whatwg.org/#concept-request-method) is not `GET`, return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a TypeError.

This means that you can't put request urls with different schemes, for example `ipfs://` in the cache, even if you have a valid constructed response for this request.

I have a program that retrieves data and caches it over both `http://` and `ipfs://`. For `ipfs://` urls, I re-write the url to fetch via a local `http://` gateway. Because I want the cache entry to be gateway-independent, I would like to store the original `ipfs://` url in the cache. However, as the spec says, I cannot do this. So the obvious solution is to just work around that by re-writing the cache url to use `http://` instead.

I do a similar thing for requests with the `Range:` header set where a partial response (status code 206) is returned. The cache normally rejects these (step 6.) so I instead resolve the response array buffer and fabricate a new response, inserting it into the cache with a `bytes=X-Y` query pair.

In summary, I think the Cache API would be much improved by removing some of these limitations so that I and others don't have to ship work-arounds to them.

**Comment by amn** [comment]
I have come to this issue searching for rationale behind not allowing HTTP method(s) other than GET (and HEAD) in request-response pairs to cache. But for a non-HTTP protocol like IPFS, what would the meaning of the e.g. `header` property on `Request` and `Response` be? Or `method`? I don't think it makes sense for the Cache API to allow something not derived from HTTP, for the reason mentioned (and related reasons).

What you're looking for is a custom cache, using a composite index modeled by a "IPFS.Request" class of your choosing, into a set of "IPFS.Response" records that conform to whatever and however resources can be represented as obtained over IPFS.

You can write an "IPFS.Cache" class that implements the `Cache` interface well enough for you to be able to use it in place of [platform] `Cache` objects.

Alternatively, if IPFS _can_ carry/work over HTTP, then I'd argue you should retain `http://` in your URLs, and bolt IPFS-specific things on top of `Cache` so it's oblivious to IPFS, while at the same time allowing extension.
